scipy/optimize/tests/test__numdiff.py

```diff
@@ -0,0 +1,275 @@
+        ub = np.array([1.0, 1.0, 1.0, 0.05])
+        h_adjusted, scheme = _adjust_step_2nd_order(x0, h, (lb, ub))
+        assert_allclose(h_adjusted, np.array([1e-1, -1e-1, 1e-1, 0.04]))
+        assert_equal(scheme, np.array([0, 1, 1, 0]))
+
+    def test_scalar(self):
+        h = np.array(0.1)
+        x0 = np.array(0.95)
+        lb = -1
+        ub = 1
+        h_adjusted, scheme = _adjust_step_2nd_order(x0, h, (None, None))
+        assert_allclose(h_adjusted, 0.1)
+        assert_equal(scheme, 0)
+        h_adjusted, scheme = _adjust_step_2nd_order(x0, h, (lb, None))
+        assert_allclose(h_adjusted, 0.1)
+        assert_equal(scheme, 0)
+        h_adjusted, scheme = _adjust_step_2nd_order(x0, h, (None, ub))
+        assert_allclose(h_adjusted, -0.1)
+        assert_equal(scheme, 1)
+        lb = -0.05
+        ub = 0.05
+        x0 = np.array(-0.01)
+        h_adjusted, scheme = _adjust_step_2nd_order(x0, h, (lb, ub))
+        assert_allclose(h_adjusted, 0.04)
+        assert_equal(scheme, 0)
+
+
+class TestApproxDerivatives(object):
+    def fun_scalar_scalar(self, x):
+        return np.sinh(x)
+
+    def jac_scalar_scalar(self, x):
+        return np.cosh(x)
+
+    def fun_scalar_vector(self, x):
+        return np.array([x**2, np.tan(x), np.exp(x)])
+
+    def jac_scalar_vector(self, x):
+        return np.array([2 * x, np.cos(x) ** -2, np.exp(x)])
+
+    def fun_vector_scalar(self, x):
+        return np.sin(x[0] * x[1]) * np.log(x[0])
+
+    def jac_vector_scalar(self, x):
+        return np.array([
+            x[1] * np.cos(x[0] * x[1]) * np.log(x[0]) +
+            np.sin(x[0] * x[1]) / x[0],
+            x[0] * np.cos(x[0] * x[1]) * np.log(x[0])
+        ])
+
+    def fun_vector_vector(self, x):
+        return np.array([
+            x[0] * np.sin(x[1]),
+            x[1] * np.cos(x[0]),
+            x[0] ** 3 * x[1] ** -0.5
+        ])
+
+    def jac_vector_vector(self, x):
+        return np.array([
+            [np.sin(x[1]), x[0] * np.cos(x[1])],
+            [-x[1] * np.sin(x[0]), np.cos(x[0])],
+            [3 * x[0] ** 2 * x[1] ** -0.5, -0.5 * x[0] ** 3 * x[1] ** -1.5]
+        ])
+
+    def fun_parametrized(self, x, c0, c1):
+        return np.array([np.exp(c0 * x[0]), np.exp(c1 * x[1])])
+
+    def jac_parametrized(self, x, c0, c1):
+        return np.array([
+            [c0 * np.exp(c0 * x[0]), 0],
+            [0, c1 * np.exp(c1 * x[1])]
+        ])
+
+    def fun_zero_jacobian(self, x):
+        return np.array([x[0] * x[1], np.cos(x[0] * x[1])])
+
+    def jac_zero_jacobian(self, x):
+        return np.array([
+            [x[1], x[0]],
+            [-x[1] * np.sin(x[0] * x[1]), -x[0] * np.sin(x[0] * x[1])]
+        ])
+
+    def test_scalar_scalar(self):
+        x0 = 1.0
+        jac_diff_f = approx_derivatives(self.fun_scalar_scalar, x0,
+                                        method='forward')
+        jac_diff_c = approx_derivatives(self.fun_scalar_scalar, x0)
```

Comment:
Might want to convert this (and similar ones below) into nose test generators for clarity.  See the 'How do I write tests with test generators?' section in http://docs.scipy.org/doc/scipy-dev/reference/hacking.html